apm-agent-plugins/apm-scala-concurrent-plugin/src/test/scala/co/elastic/apm/agent/scala/concurrent/FutureInstrumentationSpec.scala

```diff
@@ -0,0 +1,237 @@
+      true
+    )
+
+
+  }
+
+  test("Worker thread should correctly set context on the current transaction") {
+    implicit val executionContext: ExecutionContextExecutor =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(1))
+
+    new TestFutureTraceMethods().invokeAsync(tracer)
+    transaction.deactivate().end()
+    assertEquals(reporter.getTransactions.size(), 1)
+    assertEquals(reporter.getSpans.size(), 0)
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future").asInstanceOf[Boolean],
+      true
+    )
+  }
+
+  test("Multiple async operations should be able to set context on the current transaction") {
+
+    implicit val multiPoolEc: ExecutionContextExecutor =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(3))
+
+    val future = Future
+      .traverse(1 to 100) { _ =>
+        Future.sequence(List(
+          Future {
+            Thread.sleep(25)
+            tracer.currentTransaction().addCustomContext("future1", true)
+          },
+          Future {
+            Thread.sleep(50)
+            tracer.currentTransaction().addCustomContext("future2", true)
+          },
+          Future {
+            Thread.sleep(10)
+            tracer.currentTransaction().addCustomContext("future3", true)
+          }
+        ))
+      }
+
+    Await.ready(future, 10.seconds)
+    transaction.deactivate().end()
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future1").asInstanceOf[Boolean],
+      true
+    )
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future2").asInstanceOf[Boolean],
+      true
+    )
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future3").asInstanceOf[Boolean],
+      true
+    )
+
+  }
+
+  test("Handle a combination of Promises and Futures correctly") {
+
+    implicit val multiPoolEc: ExecutionContextExecutor =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(3))
+
+    val promise = Promise[Int]()
+
+    Future { Thread.sleep(100) }
+      .map(_ => 42)
+      .onComplete {
+        case Success(value) => promise.success(value)
+        case Failure(exception) => promise.failure(exception)
+      }
+
+    val future = promise
+      .future
+      .map(_ => tracer.currentTransaction().addCustomContext("future", true))
+
+    Await.ready(future, 10.seconds)
+    transaction.deactivate().end()
+    assertEquals(
+      reporter.getTransactions.get(0).getContext.getCustom("future").asInstanceOf[Boolean],
+      true
+    )
+
+  }
+
+  test("Handle a Future.sequence correctly") {
+
+    implicit val multiPoolEc: ExecutionContextExecutor =
+      ExecutionContext.fromExecutor(Executors.newFixedThreadPool(3))
+
+    val future = Future
+      .sequence(List(
+        Future(Thread.sleep(25))
+      ))
+      .map(_ => tracer.currentTransaction().addCustomContext("future", true))
```

Comment:
From what I tested with a Java Executor only:
1. If the Thread.sleep(x) with x = 1, the test does not always fail
2. Adding a map with adding customContext on the Future inside of the Sequence, the context is correctly propagated
3. So only the map after the Sequence seems to lose context when only using the Java Executor

Sequence uses the parasitic Executor from Scala.